Refactor kube-proxy service/endpoints update so that can be consumed among different proxiers

m1093782566 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

There are huge duplication among different proxiers.  For example, the service/endpoints list/watch part in iptables, ipvs and windows kernel mode(to be get in soon).

I think the more places this is replicated the harder it becomes to keep correct. We may need to refactor it and let different proxiers consume the same code.

**Which issue this PR fixes**: 

fixes #52464

**Special notes for your reviewer**:

* This refactor reduces **500** Lines in iptables proxy, so it will reduce **500*N**(number of proxiers) lines in total. People no need to care the service/endpoints update logic any more and can be more focus on proxy logic.

* I would like to do the following things in follow-ups:

1. rsync it to ipvs proxier

2. rsync it to winkernel proxier

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Refactor kube-proxy service/endpoints update so that can be consumed among different proxiers
```

m1093782566 commented:
/assign @thockin  @wojtek-t @feiskyer @dcbw

m1093782566 commented:
/assign @thockin

m1093782566 commented:
/assign @wojtek-t

m1093782566 commented:
/assign @feiskyer

m1093782566 commented:
/assign @dcbw

m1093782566 commented:
/cc @jiaxuanzhou

jiaxuanzhou commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -155,6 +153,66 @@ type serviceInfo struct {
 	serviceLBChainName       utiliptables.Chain
 }
 
+func (info *serviceInfo) ClusterIP() string {
```
@m1093782566  suggest to add some instructions of the funcs.

m1093782566 replied:
Sure, will do that.

jiaxuanzhou commented on pkg/proxy/util/endpoints.go:
```diff
@@ -0,0 +1,169 @@
+	localIPs := make(map[types.NamespacedName]sets.String)
+	for svcPortName := range endpointsMap {
+		for _, ep := range endpointsMap[svcPortName] {
+			if ep.IsLocal() {
```
how about` if !ep.IsLocal() {...continue...}` ?

thockin commented:
There's a lot here, and it is sort of scary.  I don't like the giant interface, especially since it is 1:1 (for now).  I'd rather move the struct definitions over first, which should be a relatively trivial PR, and THEN consider whether we need an interface or not.  Most of those types are just dumb data.

If we move them into a util lib, the comments on the now-public functions have to be REALLY good.  These functions were not designed to be generic - they solve a very specific purpose in a very specific context.  Making them generic means a brighter spotlight is needed on clarity, comments, and tests.

m1093782566 commented:
@thockin

I apologize that I scared you :)

Let me explain why I introduced the 2 interfaces:  `ServiceInfo` and `EndpointsInfo`. 

```
type proxyServiceMap map[proxy.ServicePortName]*serviceInfo
type proxyEndpointsMap map[proxy.ServicePortName][]*endpointsInfo
```

Currently both `serviceInfo` and `endpointsInfo` are proxy-specific, and if we keep them being proxy-specific, it's difficult to share the duplicated codes among proxiers.

I guess you may don't like defining `serviceInfo` and `endpointsInfo` in util and let all proxiers consume the same `serviceInfo` and `endpointsInfo`. Because I remember you said "When rebasing the ipvs to be the same structure as iptables, a lot of this will be different." in IPVS PR(xref https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/46580#-KrcZLOn9a9TJY6Zk1tl). So, I need a public interface.

I agree the 1:1 interface looks a bit large. So, I changed the interface definition now.

```
type ServiceInfo interface {
	ClusterIP() string
	Port() int
	Protocol() api.Protocol
	HealthCheckNodePort() int
}

type EndpointsInfo interface {
	Endpoint() string
	IsLocal() bool
	IPPart() string
	Equal(EndpointsInfo) bool
}
```

I think the structs after changing looks more clean now.

> If we move them into a util lib, the comments on the now-public functions have to be REALLY good. These functions were not designed to be generic - they solve a very specific purpose in a very specific context. Making them generic means a brighter spotlight is needed on clarity, comments, and tests.

Can't agree more. I would like to add more clarity, comments, and tests when you are green at the basic idea :) 

Please check the updates when you have a chance. Looking forwarding to hearing your opinions. Thanks!

m1093782566 commented:
/retest

m1093782566 commented:
@wojtek-t What's your opinion?

m1093782566 commented:
/retest

thockin commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,22 @@ type ServicePortName struct {
 	return fmt.Sprintf("%s:%s", spn.NamespacedName.String(), spn.Port)
 }
+
+type ServiceInfo interface {
```
comments comments comments, please.  Types and each method.

m1093782566 replied:
Soooory for delay, comments are added now.

m1093782566 replied:
Please don't be angry about the lack of comments, I was just not sure if you like this refactor...

thockin replied:
I still hope we can find a better name than "Info" ?  Ideas?  `ServicePort` ?  `ServicePortHandle` ?

m1093782566 replied:
Yeah, `ServicePort` seems to be better. 

It's fixed now. PTAL.

thockin commented on pkg/proxy/util/endpoints.go:
```diff
@@ -0,0 +1,169 @@
+	current  ProxyEndpointsMap
+}
+
+type EndpointsChangeMap struct {
```
comments comments

m1093782566 replied:
Sorry, comments are added now.

thockin reviewed (commented):
This is better, and overall I get it.

The utter lack of comments makes me unhappy.

The fact that it is a monolithic commit makes me unhappy, too.  I feel like we've discussed this a few times.

If I were doing this, I would do something like:
* 1 commit to define interfaces
* 1 commit to use the interfaces in-place
* 1 commit to move only the services code to shared place and update tests
* 1 commit to move only endpoints code to shared place and update tests

As it is, I won't demand you break it up this way, but PLEASE for the sake of future reviews, think of your poor reviewers?

m1093782566 commented:
> As it is, I won't demand you break it up this way, but PLEASE for the sake of future reviews, think of your poor reviewers?

Got it. I appreciate your help.

m1093782566 commented:
/retest

m1093782566 commented:
/area kube-proxy

cmluciano commented:
/retest

m1093782566 commented:
This work has been pending for a while. 

@thockin @wojtek-t  What's your opinion?

thockin commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+}
+
+// EndpointsInfo is an interface which abstracts an endpoints info.
+type EndpointsInfo interface {
```
Can we just call this `Endpoint` and rename Endpoint() to String() (and add String() to the Service interface too?

m1093782566 replied:
Good point!

BTW, there is already a `String()` in `endpointsInfo`, xref: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L187-L189

I find `endpointsInfo.String()` is never used and can we just change it from 

```
func (e *endpointsInfo) String() string {
	return fmt.Sprintf("%v", *e)
}
```

to 

```
func (e *endpointsInfo) String() string {
	return e.endpoint
}
```

???

thockin commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+	Port() int
+	// Protocol returns service protocol.
+	Protocol() api.Protocol
+	// HealthCheckNodePort returns service health check node port.
```
Is this "...if present" or is it always defined?  I forget

m1093782566 replied:
> Is this "...if present" or is it always defined? I forget

This is "...if present". The comment message changed to 

```
// HealthCheckNodePort returns service health check node port if present.  If return 0, it means not present.
```

Is that ok?

thockin reviewed (commented):
Just commit #1 here.  Would be nice to see this in use, as part of this commit, but not sure how hard that would be to extract and rebase

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+func (sm *ProxyServiceMap) Merge(other ProxyServiceMap) sets.String {
+	existingPorts := sets.NewString()
+	for svcPortName, info := range other {
+		existingPorts.Insert(svcPortName.Port)
```
This is going to produce a set like {"80", "8000", "443"} - it's not clear why this is useful or safe, given that the input is a a Map which could have multiple Services in it.  If I recall, we ensure that the map only has a single Service in it, but that's because it was internal.  Now that it is external, the assumptions have to be reduced, or at least made WAY more obvious.

m1093782566 replied:
> This is going to produce a set like {"80", "8000", "443"} 

No. `svcPortName.Port` is `string` type instead of `int`.

m1093782566 replied:
> t's not clear why this is useful or safe, given that the input is a a Map which could have multiple Services in it. If I recall, we ensure that the map only has a single Service in it, but that's because it was internal. Now that it is external, the assumptions have to be reduced, or at least made WAY more obvious.

I Agree - we can change 

`existingPorts.Insert(svcPortName.Port)` to `existingPorts.Insert(svcPortName.String())`, and `svcPortName.String()` is unique across all services.

m1093782566 replied:
It's fixed now, please take a look.

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+// Unmerge deletes current ProxyServiceMap's elements which are in other ProxyEndpointsMap.
+func (sm *ProxyServiceMap) Unmerge(other ProxyServiceMap, existingPorts, staleServices sets.String) {
+	for svcPortName := range other {
+		if existingPorts.Has(svcPortName.Port) {
```
I feel like existingPorts is redundant.  This is a set of port nums/names for a give Service (see above feedback - this unstated assumption is not clear).  It's not clear if that is the ports on `sm` or on `other`, but in either case we could iterate here and eliminate the argument?

m1093782566 replied:
Nice finding!

I agree, `existingPorts` is redundant and never used in this function. We can safely eliminate  it.

m1093782566 replied:
It's fixed now. PTAL.

m1093782566 replied:
> I feel like existingPorts is redundant. ...but in either case we could iterate here and eliminate the argument?

I took another look and found `existingPorts` is not redundant.

m1093782566 replied:
`existingPorts` is used to store all the unique newly merged services' identifier and help us tell if a service is updated or deleted. I already added the comment message, PTAL.

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+// Unmerge deletes current ProxyServiceMap's elements which are in other ProxyEndpointsMap.
+func (sm *ProxyServiceMap) Unmerge(other ProxyServiceMap, existingPorts, staleServices sets.String) {
+	for svcPortName := range other {
+		if existingPorts.Has(svcPortName.Port) {
```
A one-line comment here wopuld be great

m1093782566 replied:
Sounds fair. What about 

```
// existingPorts is going to store all the unique newly merged services' identifier -ServicePortName.String().
```

It's fixed now. PTAL.

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+		if exists {
+			glog.V(1).Infof("Removing service port %q", svcPortName)
+			if info.Protocol() == api.ProtocolUDP {
+				staleServices.Insert(info.ClusterIP())
```
rename staleServices to `udpRemoved` or something?

m1093782566 replied:
What about `UDPStaled` ?

m1093782566 replied:
Or, what about `UDPStaledClusterIP`? 

```
// UDPStaledClusterIP stores all the unique stale udp cluster IP.
UDPStaledClusterIP sets.String
```

thockin replied:
"staled" isn't really a word :)

m1093782566 replied:
Oh, thanks for pointing it out, then `UDPStaleClusterIP`?

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+	// HCServices is the service map which need to be health checked.
+	HCServices map[types.NamespacedName]uint16
+	// StaleServices stores all the unique stale services.
+	StaleServices sets.String
```
rename and update comments to say UDP if this is just UDP?

m1093782566 replied:
Yes, it's just UDP. 

Done.

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+// UpdateServiceMapResult is the updated results after applying service changes.
+type UpdateServiceMapResult struct {
+	// HCServices is the service map which need to be health checked.
+	HCServices map[types.NamespacedName]uint16
```
Better name needed

thockin replied:
needs documentation - what is the uint16 here?

m1093782566 replied:
> needs documentation - what is the uint16 here?

Added the comment message: 
```
// The uint16 value of HCServices map is the service health check node port.
```

m1093782566 replied:
> Better name needed

What about `HCServices` -> `HCServiceNodePorts` ?

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+	StaleServices sets.String
+}
+
+// <serviceMap> is updated by this function (based on the given changes).
```
GoDoc comment, please

m1093782566 replied:
It's fixed now. PTAL.

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+// <serviceMap> is updated by this function (based on the given changes).
+// <changes> map is cleared after applying them.
+func UpdateServiceMap(
+	serviceMap ProxyServiceMap,
```
no need to linewrap this

m1093782566 replied:
Yes, it makes code looks cleaner.

Done.

thockin commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+		changes.lock.Lock()
+		defer changes.lock.Unlock()
+		for _, change := range changes.items {
+			existingPorts := serviceMap.Merge(change.current)
```
I said before that making this public might need to make the API better - this is pretty ugly.  Maybe we need a better way to express this?  Maybe one mthod that does this internally?  `serviceMap.Apply(change)` ?

m1093782566 replied:
refactored it to `serviceMap.Apply(change)` now, PTAL.

thockin reviewed (commented):
This is just commit #2.

Where are the tests?  As a public API, this needs solid tests.

thockin reviewed (commented):
Many of the same comments apply here as for commit #2, so I won't review this one yet

m1093782566 commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+)
+
+// ProxyServiceMap maps a service to its service info.
+type ProxyServiceMap map[proxy.ServicePortName]proxy.ServiceInfo
```
I like ServiceMap

m1093782566 replied:
Changed to `ServiceMap` now.

m1093782566 commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+// ProxyServiceMap maps a service to its service info.
+type ProxyServiceMap map[proxy.ServicePortName]proxy.ServiceInfo
+
+// endpointsChange stores both the previous and current ProxyServiceMap.
```
Fixed now.

m1093782566 commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+	change, exists := scm.items[*namespacedName]
+	if !exists {
+		change = &serviceChange{}
+		change.previous = serviceToServiceMap(previous)
```
Yes. `serviceToServiceMap()` is defined in proxier.go and I just pass the function pointer in the caller side.

m1093782566 replied:
added `serviceToServiceMap` as part of this commit now. PTAL.

m1093782566 commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+// Merge adds other ProxyEndpointsMap's elements to current ProxyEndpointsMap, if exists, update it.
+// It returns all other ProxyEndpointsMap's service ports.
```
Done. PTAL.

m1093782566 commented on pkg/proxy/util/service.go:
```diff
@@ -0,0 +1,145 @@
+	return existingPorts
+}
+
+// Unmerge deletes current ProxyServiceMap's elements which are in other ProxyEndpointsMap.
```
Yeah. It's fixed now and comment message is 

'// Unmerge deletes all other ServiceMap's elements from current ServiceMap.  The returned value of Merge() is the `existingPorts` argument of Unmerge().
// If a service identifier exists in existingPorts, it means it's an `Update` action and will skip delete, otherwise it's a `Delete` action and will delete the service from ServiceMap.
// We pass in the `UDPStaleClusterIP`argument for storing the stale udp service cluster IP and clear stale udp connection later.'

BTW, I also think
`func (sm *ServiceMap) Unmerge(other ServiceMap, existingPorts, UDPStaleClusterIP sets.String)` is a weird function signature and we can refactor the `UDPStaleClusterIP` to be the returned value later. wdyt?

Commit:
abstract proxy servicePort and endpoints

Commit:
proxy service part changes

Commit:
proxy endpoints part changes

Commit:
iptables proxier part changes

m1093782566 commented:
@thockin 

All comments are addressed now. Please take a look when you have time.

BTW, this PR can be easily conflicted, unfortunately :(

m1093782566 commented:
I assume winkernel mode proxier has huge duplication codes with iptables mode as well.

/assign @brendandburns 

for thoughts.

m1093782566 commented:
/unassign @dcbw

m1093782566 commented:
@thockin 

May I ask you continue to review this PR since it changes many files in kube-proxy and can be very easily conflicted with other PRs - resolve conflicts and keep commit history clean is not very easy...

m1093782566 commented:
/retest

rramkumar1 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+	Equal(Endpoints) bool
+}
+
+// EndpointServicePair is an Endpoint and ServicePortName pair.
```
This comment is a little redundant. We obviously know EndpointServicePair contains an Endpoint and ServicePortName just by looking at it.

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -157,6 +155,30 @@ type serviceInfo struct {
+	return info.clusterIP.String()
+}
+
+// Port is part of interface.
```
Can you change the comment to something like "Port is defined in the ServicePort interface"? This makes it clear what interface you are referring to.

m1093782566 replied:
What about "Port is part of ServicePort interface"?

m1093782566 replied:
Done

rramkumar1 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -157,6 +155,30 @@ type serviceInfo struct {
+	return info.port
+}
+
+// Protocol is part of interface.
```
Same as above.

m1093782566 replied:
Done

rramkumar1 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -157,6 +155,30 @@ type serviceInfo struct {
+	return info.protocol
+}
+
+func (info *serviceInfo) String() string {
```
Not sure if you wanted to add a comment here?

m1093782566 replied:
Yes, we need it.

m1093782566 replied:
Done

rramkumar1 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -168,11 +190,34 @@ type endpointsInfo struct {
 }
 
-// IPPart returns just the IP part of the endpoint.
+// IsLocal is part of interface.
```
Explicitly state the interface in your comment. Please do it for the other commented functions as well.

m1093782566 replied:
Done

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+	// hostname is the host where kube-proxy is running.
+	hostname string
+	// items is a service to its endpointsChange map.
+	items map[types.NamespacedName]*endpointsChange
```
To me, it is cleaner if the comment reads something like: "items maps a service to an endpointsChange". If you wanted to be thorough, you could even add a little more. Possibly something like "this is used to keep track of what endpoints changed for each service"

m1093782566 replied:
Sounds fair. It's fixed now.

rramkumar1 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+// ServiceMap maps a service to its ServicePort information.
+type ServiceMap map[ServicePortName]ServicePort
+
+// NewServiceChangeMap initialize a ServiceChangeMap
```
Nit: "initialize" -> "initializes"

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+	items map[types.NamespacedName]*serviceChange
+}
+
+// serviceChange contains all changes to services that happened since proxy rules was synced.  For a single object,
```
Nit: "...since proxy rules was synced" -> "...since proxy rules _**were**_ synced"

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+type ServiceChangeMap struct {
+	// lock protects items.
+	lock sync.Mutex
+	// items is a service to its serviceChange map.
```
See my comment for items in EndpointsChangeMap

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+	items map[types.NamespacedName]*endpointsChange
+}
+
+// endpointsChange contains all changes to endpoints that happened since proxy rules was synced.  For a single object,
```
Nit: "...since proxy rules was synced" -> "...since proxy rules _**were**_ synced"

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+
+// endpointsChange contains all changes to endpoints that happened since proxy rules was synced.  For a single object,
+// changes are accumulated, i.e. previous is state from before all of them,
+// current is state after applying all of those.
```
Nit: "i.e. previous is the state before changes, current is the state after changes"?

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+
+// serviceChange contains all changes to services that happened since proxy rules was synced.  For a single object,
+// changes are accumulated, i.e. previous is state from before all of them,
+// current is state after applying all of those.
```
Nit: "i.e. previous is the state before changes, current is the state after changes"?

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+// EndpointsMap maps a service to its endpoints.
+type EndpointsMap map[ServicePortName][]Endpoints
+
+// NewEndpointsChangeMap initialize an EndpointsChangeMap
```
Nit: "initialize" -> "initializes"

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+}
+
+// UpdateEndpointsMap updates endpointsMap base on the given changes.
+// The changes map is cleared after applying them.
```
In the comment you claim that the changes map is cleared in this function, but in reality the clearing happens in endpointsMap.apply. Can you move that portion of the comment to the apply function to make it more clear which piece is actually doing the clearing?

m1093782566 replied:
Sounds fair, it's fixed now.

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+	changes.items = make(map[types.NamespacedName]*endpointsChange)
+}
+
+// Merge adds other ProxyEndpointsMap's elements to current ProxyEndpointsMap, if exists, update it, otherwise create new one.
```
I think a much more concise comment for this would read as follows: "Merge ensures that the current EndpointsMap for the proxier is rebased onto the passed in EndpointsMap" WDYT?

m1093782566 replied:
Nice comment, will follow it.

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+	}
+}
+
+// Unmerge deletes all other EndpointsMap's elements from current EndpointsMap.
```
"Unmerge removes the service -> endpoint pairs from the current EndpointsMap which are contained in the EndpointsMap passed in" WDYT?

m1093782566 replied:
Nice comment! Will copy it :)

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+	return len(ecm.items) > 0
+}
+
+// endpointsToEndpointsMap translates single Endpoints object to proxyEndpointsMap.
```
"translates single Endpoints object to proxyEndpointsMap" -> "translates **a** single Endpoints object to an **EndpointsMap**"

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+				if localIPs[nsn] == nil {
+					localIPs[nsn] = sets.NewString()
+				}
+				localIPs[nsn].Insert(ep.IPPart()) // just the IP part
```
I think you can remove the comment here since it is just reiterating what the code is clearly doing.

m1093782566 replied:
It's fixed now.

rramkumar1 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
 	return fmt.Sprintf("%s:%s", spn.NamespacedName.String(), spn.Port)
 }
+
+// ServicePort is an interface which abstracts a service information.
```
After this change serviceInfo will not exist anymore right? I'm not sure what the best way to change this comment is but it definitely should not include reference to serviceInfo

rramkumar1 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+	HealthCheckNodePort() int
+}
+
+// Endpoints is an interface which abstracts an endpoints info.
```
Same as above.

rramkumar1 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
 	return fmt.Sprintf("%s:%s", spn.NamespacedName.String(), spn.Port)
 }
+
+// ServicePort is an interface which abstracts a service information.
```
"which abstracts service info"

m1093782566 replied:
Okay, seems more consistent with Endpoints interface. 

Done.

rramkumar1 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+	return len(scm.items) > 0
+}
+
+// serviceToServiceMap translates single Service object to proxyServiceMap.
```
"translates single Service object to proxyServiceMap" -> "translate **a** single Service object to a ServiceMap"

m1093782566 replied:
It's fixed now.

rramkumar1 reviewed (commented):
@m1093782566 Lots of nits. Please let me know if you have any questions. Will review unit tests once you get through these comments.

rramkumar1 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+}
+
+// Endpoints is an interface which abstracts an endpoints info.
+type Endpoints interface {
```
What is the purpose of Endpoints interface and ServicePort interface? I'm a little confused on what you are trying to accomplish.

m1093782566 replied:
See comments in https://github.com/kubernetes/kubernetes/issues/52464. Briefly speaking, interface `Endpoints` and `ServicePort` is used to abstract the current `serviceInfo` and `endpointsInfo`. If we keep them private variable, we are not able to do refactor.

m1093782566 replied:
After the refactor, IPVS kube-proxy will share the same public Endpoints, ServicePort interfaces and their public methods.

rramkumar1 replied:
Thanks for clarifying :)

m1093782566 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -157,6 +155,30 @@ type serviceInfo struct {
+	return fmt.Sprintf("%s:%d/%s", info.clusterIP, info.port, info.protocol)
+}
+
+// HealthCheckNodePort is part of interface.
```
Done

m1093782566 commented:
@rramkumar1 

Thanks for your diligently review and so many nice findings. They are all addressed now, PTAL. THANKS!

rramkumar1 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -182,12 +228,8 @@ func (e *endpointsInfo) endpointChain(svcNameString, protocol string) utiliptabl
-
 // returns a new serviceInfo struct
-func newServiceInfo(svcPortName proxy.ServicePortName, port *api.ServicePort, service *api.Service) *serviceInfo {
+func newServiceInfo(svcPortName proxy.ServicePortName, port *api.ServicePort, service *api.Service) proxy.ServicePort {
```
change func name to newServicePort since you are returning a proxy.ServicePort?

rramkumar1 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -182,12 +228,8 @@ func (e *endpointsInfo) endpointChain(svcNameString, protocol string) utiliptabl
-	return fmt.Sprintf("%v", *e)
-}
-
 // returns a new serviceInfo struct
```
"returns a new proxy.ServicePort"?

m1093782566 replied:
Great, will follow it!

m1093782566 replied:
Done.

rramkumar1 commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -750,150 +634,11 @@ func (proxier *Proxier) OnEndpointsSynced() {
-		servicePort := &service.Spec.Ports[i]
-		svcPortName := proxy.ServicePortName{NamespacedName: svcName, Port: servicePort.Name}
-		serviceMap[svcPortName] = newServiceInfo(svcPortName, servicePort, service)
+func newEndpointsInfo(IP string, port int, isLocal bool) proxy.Endpoints {
```
This needs a comment similar to newServiceInfo. Also should probably change the name to newEndpoints since you are now returning a proxy.Endpoints.

rramkumar1 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+	HealthCheckNodePort() int
+}
+
+// Endpoints is an interface which abstracts an endpoints info.
```
After taking a second look and considering this will be a "public" API, I think the command should read like "Endpoints in an interface which abstracts information about an endpoint". This seems more generic to me, WDYT?

m1093782566 replied:
Yeah, it's a good comment message :) Thanks! It's fixed now.

rramkumar1 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
 	return fmt.Sprintf("%s:%s", spn.NamespacedName.String(), spn.Port)
 }
+
+// ServicePort is an interface which abstracts a service info.
```
Same as below. WDYT about "ServicePort is an interface which abstracts information about a service"

m1093782566 replied:
Great! It's fixed now.

rramkumar1 commented on pkg/proxy/service_test.go:
```diff
@@ -0,0 +1,534 @@
+			},
+		},
+		{
+			// Case[7]: externalname service
```
Nit: "externalname" should be "external name"

m1093782566 replied:
nice finding! It's fixed now. PTAL. THANKS!

rramkumar1 commented on pkg/proxy/service_test.go:
```diff
@@ -0,0 +1,534 @@
+		},
+		{
+			// Case[1]: headless service
+			service: makeTestService("somewhere-else", "headless", func(svc *api.Service) {
```
Instead of using "somewhere-else" for the namespace, can you use something a little clearer like "test-ns" or "test-namespace". Just something that indicates that the argument is a namespace so it is easier to read. You can make this change in quite a few places.

rramkumar1 commented on pkg/proxy/service_test.go:
```diff
@@ -0,0 +1,534 @@
+			service: makeTestService("somewhere-else", "cluster-ip", func(svc *api.Service) {
+				svc.Spec.Type = api.ServiceTypeClusterIP
+				svc.Spec.ClusterIP = "172.16.55.4"
+				svc.Spec.Ports = addTestPort(svc.Spec.Ports, "something", "UDP", 1234, 4321, 0)
```
Same idea as above. Try giving it a name like "test-port" or "port1" or something like that just to clearly indicate what it is.

rramkumar1 commented on pkg/proxy/service_test.go:
```diff
@@ -0,0 +1,534 @@
+/*
```
General feedback about this file: The arguments you pass in for your test helper function have meaningless names most of the time (i.e "somewhere", "blah"). Can you make it so that the name of the arguments makes it clear what it is? My comments below have some examples. Please try to be consistent with how you name things as well. (i.e if you pass in "test-ns" for a namespace, do it everywhere.

m1093782566 replied:
Done. PTAL, THANKS!

rramkumar1 reviewed (commented):
Last round of feedback done. This LGTM once these final changes are in. I think the next step is to have IPVS use these interfaces & structs?

m1093782566 commented:
>  I think the next step is to have IPVS use these interfaces & structs?
Yeah, I agree. Will add it. Thanks!

rramkumar1 commented:
@m1093782566 Looks like you need to run gofmt

m1093782566 commented:
Oops.. I will do that and will push an update on IPVS proxier. Will let you know when ready. Thanks @rramkumar1.

rramkumar1 commented:
@m1093782566 Can we do the ipvs changes in a seperate PR? It would be nice to get this in first.

m1093782566 commented:
@rramkumar1 I pushed an update to fix gofmt error.

Yeah, I agree we can do ipvs part changes in another PR for the sake of soak it in CI. 

Please take a look again when you have a chance, Thanks!

rramkumar1 commented:
ping @thockin for approval

thockin commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// Unmerge deletes all other ServiceMap's elements from current ServiceMap.  The returned value of Merge() is the `existingPorts` argument of Unmerge().
+// If a service's identifier exists in `existingPorts`, it means it's an `Update` event and will skip delete, otherwise it's a `Delete` event and will
```
existingPorts is an example of a weird API that is OK privately but less so publicly.

Can we change this API so that the caller has to pre-filter?  It looks like it is only called from one place, so just adding a ServiceMap.Filter(ports sets.String) or even changing merge to return a ServiceMap instead of string set?

I could take this as a followup if needed, but please think about it.  It's just a bad API to expose.  Weird.

m1093782566 replied:
Okay, I will think about it.

m1093782566 replied:
I fixed it by adding a `ServiceMap.filter(ports sets.String)` before calling unmerge. 

BTW, I think we take `merge()`, `unmerge()` and `filter()` as private.

Please take a look to see if you are happy with these changes :)

thockin commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+	// IsLocal returns if endpoints is local.
+	IsLocal() bool
+	// IPPart returns IP part of endpoints.
+	IPPart() string
```
or Address() (ip string, port int) ?

m1093782566 replied:
I am going to change `IPPart()` to `IP()` for now. We can revise it later...

thockin commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+)
+
+// EndpointsChangeMap keeps track of endpoint changes for services.
+type EndpointsChangeMap struct {
```
Same as Service commit naming - EndpointChangeTracker ?

m1093782566 replied:
Done.

thockin commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+type EndpointsMap map[ServicePortName][]Endpoints
+
+// NewEndpointsChangeMap initializes an EndpointsChangeMap
+func NewEndpointsChangeMap(hostname string) EndpointsChangeMap {
```
Same comment as Service commit - in general, most readable code layout is:

1) type definition

2) constructor function(s)

3) methods


please move things around to this style - easier than jumping around and searching

m1093782566 replied:
Done. I think it looks better now.

thockin commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+	StaleServiceNames map[ServicePortName]bool
+}
+
+// UpdateEndpointsMap updates endpointsMap base on the given changes.
```
document the hostname argument

m1093782566 replied:
argument hostname is never used. We can remove it safely.

thockin commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,231 @@
+	result.StaleEndpoints = make(map[EndpointServicePair]bool)
+	result.StaleServiceNames = make(map[ServicePortName]bool)
+
+	func() {
```
why is this in a func() ?  comment or remove

m1093782566 replied:
It's a legacy code, we can remove it safely.

thockin commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -157,6 +155,31 @@ type serviceInfo struct {
+func (info *serviceInfo) HealthCheckNodePort() int {
+	return info.healthCheckNodePort
+}
+
```
add a type assert:

`var _ proxy.ServicePort = &serviceInfo{}` so it will fail if interface changes.  Same for Endpoint

m1093782566 replied:
Done.

thockin reviewed (commented):
I reviewed commits in order.  I got less rigorous as it went on.  In general, all the naming and style comments on services apply to endpoints.

The comments on this are MUCH nicer, so thanks for the progress so far.

m1093782566 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+}
+
+// EndpointServicePair is used to identify a service and one of its endpoint pair.
+type EndpointServicePair struct {
```
Done.

m1093782566 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+type Endpoints interface {
+	// String returns endpoints string.
+	String() string
+	// IsLocal returns if endpoints is local.
```
`IsLocal returns if endpoints is local - local means running in same host as kube-proxy.`

It's fixed now.

m1093782566 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+// Endpoints in an interface which abstracts information about an endpoint.
+type Endpoints interface {
+	// String returns endpoints string.
+	String() string
```
string format should be `IP:Port`

We also take the returned value of String() as ServiceEndpoint.Endpoint.

Done.

m1093782566 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+}
+
+// Endpoints in an interface which abstracts information about an endpoint.
+type Endpoints interface {
```
Oops, I will check the singular and plural everywhere. It should be singular here, 

```
// Endpoints in an interface which abstracts information about an endpoint.
type Endpoint interface {

}
```

m1093782566 replied:
It's fixed now, PTAL.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+	return len(scm.items) > 0
+}
+
+// serviceToServiceMap translates a single Service object to a proxyServiceMap.
```
Nice finding. Done.

m1093782566 commented on pkg/proxy/types.go:
```diff
@@ -42,3 +43,33 @@ type ServicePortName struct {
+
+// ServicePort is an interface which abstracts information about a service.
+type ServicePort interface {
+	// String returns service string.
```
Done.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+//   - pass <oldService, service> as the <previous, current> pair.
+// Delete item
+//   - pass <service, nil> as the <previous, current> pair.
+func (scm *ServiceChangeMap) Update(namespacedName *types.NamespacedName, previous, current *api.Service, makeServicePort func(svcPortName ServicePortName, servicePort *api.ServicePort, service *api.Service) ServicePort) bool {
```
> Can we please put methods for a type right next to the type definition?

Done. Please take a look again if you are happy with it...

> types.NamespacedName probably should never be a pointer

Agree, it's fixed now.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+// Merge adds other ServiceMap's elements to current ServiceMap.  If exists, update the existing one, otherwise create new one.
+// It returns a string type set which stores all the newly merged services' identifier - ServicePortName.String() to help users
+// tell if a service is deleted or updated.  The returned value is one of the arguments of Unmerge().  For example,
+// - assume serviceMap A{} want to merge serviceMap B{{"ns", "cluster-ip", "http"}: {"172.16.55.10", 1234, "TCP"}}, then it will
```
Okay, I optimized the example format.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+
+// Merge adds other ServiceMap's elements to current ServiceMap.  If exists, update the existing one, otherwise create new one.
+// It returns a string type set which stores all the newly merged services' identifier - ServicePortName.String() to help users
+// tell if a service is deleted or updated.  The returned value is one of the arguments of Unmerge().  For example,
```
Done.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+// UpdateServiceMap updates ServiceMap based on the given changes.
+func UpdateServiceMap(serviceMap ServiceMap, changes *ServiceChangeMap) (result UpdateServiceMapResult) {
+	result.UDPStaleClusterIP = sets.NewString()
+	func() {
```
It's a legacy code... seems no special meaning.

I think we can remove it safely.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+	// HCServices is the service map which need to be health checked.
+	// The value(uint16) of HCServices map is the service health check node port.
+	HCServiceNodePorts map[types.NamespacedName]uint16
+	// UDPStaleClusterIP stores all the unique stale udp cluster IP.
```
Good comment message, will copy it :)

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+
+// UpdateServiceMapResult is the updated results after applying service changes.
+type UpdateServiceMapResult struct {
+	// HCServices is the service map which need to be health checked.
```
> comment and code don't match - field names

Oops... It's fixed now.

Another nice comment, going to copy it :)

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+type ServiceMap map[ServicePortName]ServicePort
+
+// NewServiceChangeMap initializes a ServiceChangeMap
+func NewServiceChangeMap() ServiceChangeMap {
```
Done.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+)
+
+// ServiceChangeMap keeps track of service changes.
+type ServiceChangeMap struct {
```
Agree, ServiceChangeTracker is a more generic name.

It's fixed now.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+	proxyutil "k8s.io/kubernetes/pkg/proxy/util"
+)
+
+// ServiceChangeMap keeps track of service changes.
```
Done. Thanks!

Commit:
ipvs part changes

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,193 @@
+}
+
+// UpdateServiceMap updates ServiceMap based on the given changes.
+func UpdateServiceMap(serviceMap ServiceMap, changes *ServiceChangeMap) (result UpdateServiceMapResult) {
```
Yeah, I append the IPVS part changes in this PR - one commit after iptables part changes.

I will send winkernel part changes later, but I think it's suffie to see the impact on caller sides?

m1093782566 commented:
@thockin 

I am happy that keeping commit history clean helped review :)

Comments are addressed now, please take a look again when you have a chance.

BTW, I also append a commit for IPVS part changes.

thockin commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+// GetLocalIPs returns endpoints IPs if given endpoints is local - local means the endpoint is running in same host as kube-proxy.
+func GetLocalIPs(endpointsMap EndpointsMap) map[types.NamespacedName]sets.String {
```
should be GetLocalEndpointIPs ?

m1093782566 replied:
Yeah, seems `GetLocalEndpointIPs` is more descriptive.

Done.

thockin reviewed (commented):
Small comments.  Approving.  Please fixup.  If you beat the merge queue, ping me to re-lgtm

thockin commented:
/lgtm
/approve

If you don't mind, I am going to use this PR as a basis for a code-review tutorial.  It's big enough and complex enough, but it turned out pretty nice.  I'll actually dirty it up so I have something good to point out during the session :)

m1093782566 commented:
/hold

I am happy that this PR can be a code review tutorial material, so I will continue to fix review comments util it looks pretty good :)

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,210 @@
+//   - pass <oldService, service> as the <previous, current> pair.
+// Delete item
+//   - pass <service, nil> as the <previous, current> pair.
+func (sct *ServiceChangeTracker) Update(namespacedName types.NamespacedName, previous, current *api.Service, makeServicePort func(svcPortName ServicePortName, servicePort *api.ServicePort, service *api.Service) ServicePort) bool {
```
> For cleanup: Do we need the namespacedName argument? Or can we extract it from the previous/current?

Good point! We can extract it by doing

```
endpoints := current
	if endpoints == nil {
		endpoints = previous
	}
	namespacedName := types.NamespacedName{Namespace: endpoints.Namespace, Name: endpoints.Name}
```



It's fixed now. PTAL.

m1093782566 replied:
> We need a comment about what makeServicePort() does. I can't see from this PR how it is used, and it is not part of the original APIs.

// makeServicePort() return a proxy.ServicePort based on the given Service and its ServicePort.  We inject makeServicePort()
// so that giving caller side a chance to initialize proxy.ServicePort interface.

m1093782566 replied:
Done. PTAL.

m1093782566 commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,210 @@
+
+// serviceToServiceMap translates a single Service object to a ServiceMap.
+//
+// NOTE: service object should NOT be modified.
```
// makeServicePort() return a proxy.ServicePort based on the given Service and its ServicePort.  We inject makeServicePort()
// so that giving caller side a chance to initialize proxy.ServicePort interface.

Done. PTAL.

m1093782566 commented:
/hold cancel

I apologize I didn't let the comments on `IsLocal() bool` in

```
// IsLocal returns true if the endpoint is running in same host as kube-proxy, otherwise returns false.
```

Probably I lost it during rebase.

All comments are addressed. Please take a look again :)

m1093782566 commented:
/retest

m1093782566 commented:
friendly ping @thockin for re-lgtm :)

prameshj commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,236 @@
+			if ep.IsLocal() {
+				nsn := svcPortName.NamespacedName
+				if localIPs[nsn] == nil {
+					localIPs[nsn] = sets.NewString()
```
Since the only usage of GetLocalEndpointIPs is using this map to calculate the count, should this function just return a map[types.NamespacedName]int which contains the count directly?

m1093782566 replied:
Of course we can do that, but I can not see any significant benefit. It's just a matter of counting inside vs. outside.

I would like to keep as it is.

m1093782566 replied:
We still need sets.NewString() to filter out unique ep IPs - just for clarifying :)

prameshj commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1164,7 +901,12 @@ func (proxier *Proxier) syncProxyRules() {
+	for svcName, svc := range proxier.serviceMap {
+		svcInfo, ok := svc.(*serviceInfo)
+		if !ok {
+			glog.Errorf("Failed to cast serviceInfo")
```
Would be useful to include svcName in this log message.

m1093782566 replied:
Make sense, done.

prameshj commented on pkg/proxy/iptables/proxier.go:
```diff
@@ -1448,8 +1190,13 @@ func (proxier *Proxier) syncProxyRules() {
-			endpointChain = ep.endpointChain(svcNameString, protocol)
+			epInfo, ok := ep.(*endpointsInfo)
+			if !ok {
+				glog.Errorf("Failed to cast ep to endpointsInfo")
```
Maybe include svcName and ep in the log message?

m1093782566 replied:
Done.

prameshj commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -1090,15 +850,25 @@ func (proxier *Proxier) syncProxyRules() {
+	for svcName, svc := range proxier.serviceMap {
+		svcInfo, ok := svc.(*serviceInfo)
+		if !ok {
+			glog.Errorf("Failed to cast serviceInfo")
```
Would be useful to include svcName in the log message

m1093782566 replied:
Done.

prameshj commented on pkg/proxy/ipvs/proxier.go:
```diff
@@ -1090,15 +850,25 @@ func (proxier *Proxier) syncProxyRules() {
+		for _, e := range proxier.endpointsMap[svcName] {
+			ep, ok := e.(*endpointsInfo)
+			if !ok {
+				glog.Errorf("Failed to cast ep to endpointsInfo")
```
Maybe include svcName and e in the log message?

m1093782566 replied:
Done.

prameshj commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,221 @@
+	current  ServiceMap
+}
+
+// ServiceChangeTracker carries state about uncommitted changes to an arbitrary number of
```
Could you explain why this says "uncommitted changes"? I thought serviceChange holds previous and current mapping after changes have been applied. I think I am missing something.

m1093782566 replied:
```
// serviceChange contains all changes to services that happened since proxy rules were synced.  For a single object,
// changes are accumulated, i.e. previous is state from before applying the changes,
// current is state after applying all of the changes.
type serviceChange struct {
	previous ServiceMap
	current  ServiceMap
}
```

So, I think "uncommitted changes" is accurate?

prameshj replied:
ok, I think i get it now.

prameshj commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,221 @@
+	svc := current
+	if svc == nil {
+		svc = previous
+	}
```
nit: maybe check if svc is still nil, in case previous and current were incorrectly passed in as nil?

m1093782566 replied:
Done.

prameshj commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,221 @@
+//     - A updated to be {{"ns", "cluster-ip", "http"}: {"172.16.55.10", 1234, "TCP"}}
+//     - produce string set {"ns/cluster-ip:http"}
+func (sm *ServiceMap) merge(other ServiceMap) sets.String {
+	// existingPorts is going to store all the unique newly merged services' identifier.
```
looks like existingPorts contains identifiers of all services in 'others' map, not just the newly merged ones right? Should we fix the comment?

m1093782566 replied:
Nice finding, it's fixed now.

prameshj commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,221 @@
+}
+
+// filter filters out elements from ServiceMap base on given ports string sets.
+func (sm *ServiceMap) filter(ports sets.String) {
```
Any reason we can't pass a ServiceMap to the filter function? Like - func (sm *ServiceMap) filter(other *ServiceMap). 
If so, we can use that in line 153 as change.previous.filter(changes.current)
Then we don't need this existingPorts set to be returned by merge().

m1093782566 replied:
Good point, we can clean up it.

It's fixed now. PTAL. Thanks!

prameshj commented on pkg/proxy/service.go:
```diff
@@ -0,0 +1,221 @@
+			}
+			delete(*sm, svcPortName)
+		} else {
+			glog.Errorf("Service port %q removed, but doesn't exists", svcPortName)
```
Looks like service port is not being removed in this case, maybe change the log to say "Service port %q does not exist"?

m1093782566 replied:
make sense, it's fixed now.

prameshj reviewed (commented):
I just started getting familiar with the ipvs and iptables proxier code. So, apologies if I misunderstood the implementation. I have a few comments/questions.

m1093782566 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,236 @@
+	endpoints := current
+	if endpoints == nil {
+		endpoints = previous
+	}
```
Make sense, previous == nil && current == nil is unexpected, and we should return false directly.

m1093782566 replied:
It's fixed now. Thanks!

m1093782566 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,236 @@
+type UpdateEndpointMapResult struct {
+	// HCEndpointsLocalIPSize maps an endpoints name to the length of its local IPs.
+	HCEndpointsLocalIPSize map[types.NamespacedName]int
+	// StaleServiceNames identifies if an endpoints service pair is stale.
```
nice finding, it's fixed now. Thanks!

m1093782566 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,236 @@
+// is passed in to store the stale udp endpoints and `staleServiceNames` argument is passed in to store the stale udp service.
+// The changes map is cleared after applying them.
+func (endpointsMap EndpointsMap) apply(changes *EndpointChangeTracker, staleEndpoints map[ServiceEndpoint]bool, staleServiceNames map[ServicePortName]bool) {
+	changes.lock.Lock()
```
Done.

m1093782566 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,236 @@
+// GetLocalEndpointIPs returns endpoints IPs if given endpoints is local - local means the endpoint is running in same host as kube-proxy.
+func GetLocalEndpointIPs(endpointsMap EndpointsMap) map[types.NamespacedName]sets.String {
+	localIPs := make(map[types.NamespacedName]sets.String)
+	for svcPortName := range endpointsMap {
```
It's fixed now.

m1093782566 commented on pkg/proxy/endpoints.go:
```diff
@@ -0,0 +1,236 @@
+	// StaleServiceNames identifies if an endpoints service pair is stale.
+	StaleEndpoints map[ServiceEndpoint]bool
+	// StaleServiceNames identifies if a service is stale.
+	StaleServiceNames map[ServicePortName]bool
```
Make sense, it's fixed now!

Commit:
update bazel BUILD

m1093782566 commented:
@prameshj 

Thanks for so many good points, it's fixed now. PTAL. Thanks!

prameshj commented:
Thanks for addressing all the comments!

m1093782566 commented:
Kindly ping @thockin  for re-lgtm.

I am going to be out of office for Chinese Spring Festival. I think it's better to soak it in test earlier.

@rramkumar1 

Would you try lgtm :)

m1093782566 commented:
I am going to self lgtm to respect code freeze - I will be back to office in 1 or 2 weeks

/lgtm

m1093782566 commented:
Well...

thockin commented:
/lgtm
